### PR TITLE
Gsdx: Add Mipmap (Software mode) and preload data frame to GUI.

### PR DESCRIPTION
**Pull request by ssakash**
~~Windows build for testing : https://www.dropbox.com/s/o4xb7n2z1fo6a1s/GSdx32-SSE2.dll?dl=0~~ outdated build of the commit.

### CODE REVIEW

**Comment by willkuer** [comment]
This is only for troubleshooting, or? Does mipmapping has negative effects?

**Comment by ssakash** [comment]
I haven't observed any flaws of Mipmapping on software mode. the intent of the code is for easier debugging and to also keep up consistency with Linux GUI which also has option for mipmap.

File: plugins/GSdx/GSRenderer.cpp [hunk 1]
```diff
@@ -574,11 +574,13 @@ void GSRenderer::KeyEvent(GSKeyEventData* e)
 			return;
 		case VK_DELETE:
 			m_aa1 = !m_aa1;
-			printf("GSdx: (Software) aa1 is now %s.\n", m_aa1 ? "enabled" : "disabled");
+			theApp.SetConfig("aa1", m_aa1);
```
Comment (gregory38) [inline plugins/GSdx/GSRenderer.cpp]: Wasn't hotkey change suppose to be temporary?

File: plugins/GSdx/GSSetting.cpp [hunk 2]
```diff
@@ -124,10 +124,10 @@ const char* dialog_message(int ID, bool* updateText) {
 		case IDC_LOGZ:
 			return "Treat depth as logarithmic instead of linear. Recommended setting is on unless it causes graphical glitches.";
 #endif
-#ifdef __linux__
 		case IDC_MIPMAP:
 			return "Enables mipmapping, which some games require to render correctly. Turn off only for debug purposes.";
-#endif
+		case IDC_PRELOAD_GS:
+			return "The Following Hack uploads GS data when rendering a new frame to reprdouce some effects accurately, Fixes Black Screen issues in games like Primal. ";
```
Comment (gregory38) [inline plugins/GSdx/GSSetting.cpp]: Search in the issue report the other game that require the option.

File: plugins/GSdx/GSSettingsDlg.cpp [hunk 3]
```diff
@@ -423,6 +430,7 @@ void GSSettingsDlg::UpdateControls()
 		EnableWindow(GetDlgItem(m_hWnd, IDC_UPSCALE_MULTIPLIER_TEXT), hw);
 		EnableWindow(GetDlgItem(m_hWnd, IDC_FILTER), hw);
 		EnableWindow(GetDlgItem(m_hWnd, IDC_PALTEX), hw);
+		EnableWindow(GetDlgItem(m_hWnd, IDC_PRELOAD_GS), hw);
```
Comment (gregory38) [inline plugins/GSdx/GSSettingsDlg.cpp]: For me it must go into the hack section.

File: plugins/GSdx/GSRenderer.cpp (hunk 1 above)
Comment (ssakash) [inline plugins/GSdx/GSRenderer.cpp]: usually, but I don't think there are any benefits in a temporary change. users couldn't find whether currently Mipmap is enabled (or) not after using multiple toggles without the following change. ( for those who don't use the console log). This change is currently important due to fact that users might take a look at the GUI for current status of Mipmap and AA1, outdated value might confuse them.

File: plugins/GSdx/GSSettingsDlg.cpp (hunk 3 above)
Comment (ssakash) [inline plugins/GSdx/GSSettingsDlg.cpp]: I'll move it to hacks dialog , my fault. :p

File: plugins/GSdx/GSSetting.cpp (hunk 2 above)
Comment (ssakash) [inline plugins/GSdx/GSSetting.cpp]: Added Armored core: Last Raven to the tool tip instead of Primal, since Primal was only fixed on replay of dump while using ``Preload Data Frame`` not while playing the game itself.

**Comment by ssakash** [comment]
**Review of updated change with respect to master branch**

* added option for preload data frame to hacks dialog.

* made mipmap and AA1 hotkey to do a permanent change similar to shader toggle hotkey for preventing any GUI conflicts.

*added option for enabling/disabling mipmap on software mode options groupbox.

File: plugins/GSdx/GSSetting.cpp [hunk 4]
```diff
@@ -124,10 +124,10 @@ const char* dialog_message(int ID, bool* updateText) {
 		case IDC_LOGZ:
 			return "Treat depth as logarithmic instead of linear. Recommended setting is on unless it causes graphical glitches.";
 #endif
-#ifdef __linux__
 		case IDC_MIPMAP:
 			return "Enables mipmapping, which some games require to render correctly. Turn off only for debug purposes.";
-#endif
+		case IDC_PRELOAD_GS:
+			return "The Following Hack uploads GS data when rendering a new frame to reprdouce some effects accurately, Fixes Black Screen issues in games like Armored Core: Last Raven. ";
```
Comment (turtleli) [inline plugins/GSdx/GSSetting.cpp]: Either make IDC_PRELOAD_GS Windows only, or add an enum for Linux, otherwise it'll break Linux compilation.

The tooltip string needs fixed. "The Following Hack" isn't needed, reproduce is misspelt, comma should be full stop, black screen does not need capitals, and there's a space at the end.

File: plugins/GSdx/GSdx.rc [hunk 5]
```diff
@@ -213,31 +214,32 @@ BEGIN
     LTEXT           "CRC Hack Level:",IDC_CRC_LEVEL_TEXT,22,208,79,8
     COMBOBOX        IDC_CRC_LEVEL,105,206,127,63,CBS_DROPDOWNLIST | WS_VSCROLL | WS_TABSTOP
     CONTROL         "Enable HW Hacks",IDC_HACKS_ENABLED,"Button",BS_AUTOCHECKBOX | WS_TABSTOP,11,223,71,10
-    PUSHBUTTON      "Configure Hacks...",IDC_HACKSBUTTON,105,221,127,14
+    PUSHBUTTON      "Configure Hacks",IDC_HACKSBUTTON,105,221,127,14
```
Comment (turtleli) [inline plugins/GSdx/GSdx.rc]: This makes it inconsistent with the "Configure Shade Boost..." button. Either undo this change or change the shade boost button as well.

Also, might be a good idea to provide screenshots of the dialogs.

Comment (ssakash) [inline plugins/GSdx/GSdx.rc]: I remove the ... stuffs since it's just a waste of space, can't really post the screenshots right now since I did these changes on Git online editor through my mobile. would be nice if you could double check the recent changes :)

File: plugins/GSdx/GSSetting.cpp (hunk 4 above)
Comment (ssakash) [inline plugins/GSdx/GSSetting.cpp]: added it to the previous #ifdef  :heavy_check_mark:

**Comment by gregory38** [comment]
We really need to clarify the permanent/temporary option when it is changed by hotkey.
@bositman, please give your opinion on the 2nd commit.Thanks

**Comment by bositman** [comment]
I think the hotkey change should not be saved in the ini. It's an option that rarely gets toggled and when it does it's almost always to debug sutff / check for issues. Thus keep the hotkey as a temporary toggle and not permanent

**Comment by ssakash** [comment]
modified second commit to only deal with the console log message changes. (removed the permanent change stuffs) 

but in my humble opinion, temporary change doesn't give any benefits over permanent  change. the former only misleads users due to the outdated GUI status of the options.

File: plugins/GSdx/GSSetting.cpp [hunk 6]
```diff
@@ -123,11 +123,12 @@ const char* dialog_message(int ID, bool* updateText) {
 			return "Makes textures partially or fully transparent as required by emulation. May cause unusual slowdowns for some games.";
 		case IDC_LOGZ:
 			return "Treat depth as logarithmic instead of linear. Recommended setting is on unless it causes graphical glitches.";
+		// Exclusive for Hardware Renderer	
+		case IDC_PRELOAD_GS:
+			return " Uploads GS data when rendering a new frame to reproduce some effects accurately. Fixes Black Screen issues in games like Primal.";
```
Comment (turtleli) [inline plugins/GSdx/GSSetting.cpp]: Black Screen -> black screen

File: plugins/GSdx/GSRenderer.cpp [hunk 7]
```diff
@@ -574,11 +574,11 @@ void GSRenderer::KeyEvent(GSKeyEventData* e)
 			return;
 		case VK_DELETE:
 			m_aa1 = !m_aa1;
-			printf("GSdx: (Software) aa1 is now %s.\n", m_aa1 ? "enabled" : "disabled");
+			printf("GSdx: (Software) Edge Anti-aliasing is now %s.\n", m_aa1 ? "enabled" : "disabled");
```
Comment (turtleli) [inline plugins/GSdx/GSRenderer.cpp]: "Anti" IMO should be "anti"

You should probably make the Linux and Windows messages the same as well (including whether they are printed to stdout or stderr)

**Comment by turtleli** [comment]
Maybe the best thing to do is to provide an option on whether the toggle is temporary or permanent. That would satisfy everyone - or not, because then people would argue about what the default behaviour should be.

**Comment by ssakash** [comment]
> Maybe the best thing to do is to provide an option on whether the toggle is temporary or permanent. That would satisfy everyone - **or not, because then people would argue about what the default behaviour should be**.

you got the point. I might consider opening an issue/poll later for knowing user preference on either temporary / permanent change.

Fixed those other concerns on the toggle messages and uppercase -> lowercase stuffs. Thanks

File: plugins/GSdx/GSRenderer.cpp [hunk 8]
```diff
@@ -610,19 +610,19 @@ void GSRenderer::KeyEvent(GSKeyEventData* e)
...
 			return;
 		case XK_Insert:
 			m_mipmap = !m_mipmap;
-			fprintf(stderr,"GSdx: (Software) mipmapping is now %s.\n", m_mipmap ? "enabled" : "disabled");
+			printf("GSdx: (Software) Mipmapping is now %s.\n", m_mipmap ? "enabled" : "disabled");
 			return;
 		case XK_Prior:
 			m_fxaa = !m_fxaa;
-			fprintf(stderr,"GSdx: fxaa is now %s.\n", m_fxaa ? "enabled" : "disabled");
+			printf("GSdx: fxaa is now %s.\n", m_fxaa ? "enabled" : "disabled");
```
Comment (turtleli) [inline plugins/GSdx/GSRenderer.cpp]: Use the same message as Windows here?

File: plugins/GSdx/GSSetting.cpp [hunk 9]
```diff
@@ -123,11 +123,12 @@ const char* dialog_message(int ID, bool* updateText) {
 			return "Makes textures partially or fully transparent as required by emulation. May cause unusual slowdowns for some games.";
 		case IDC_LOGZ:
 			return "Treat depth as logarithmic instead of linear. Recommended setting is on unless it causes graphical glitches.";
+		// Exclusive for Hardware Renderer	
+		case IDC_PRELOAD_GS:
+			return " Uploads GS data when rendering a new frame to reproduce some effects accurately. Fixes black Screen issues in games like Primal.";
```
Comment (turtleli) [inline plugins/GSdx/GSSetting.cpp]: Screen -> screen

**Comment by ssakash** [comment]
- changed Linux FXAA toggle message to be similar to windows one.
- Screen -> screen

File: plugins/GSdx/GSdx.rc [hunk 10]
```diff
@@ -89,13 +89,14 @@ IDB_LOGOGL              BITMAP                  "res\\logo-ogl.bmp"
...
+IDD_HACKS DIALOGEX 0, 0, 147, 200
 STYLE DS_SETFONT | DS_MODALFRAME | DS_FIXEDSYS | WS_POPUP | WS_CAPTION | WS_SYSMENU
 CAPTION "Hacks Configuration"
 FONT 8, "MS Shell Dlg", 400, 0, 0x1
 BEGIN
-    DEFPUSHBUTTON   "OK",IDOK,84,163,56,14
-    GROUPBOX        "USE AT YOUR OWN RISK!",IDC_STATIC,7,7,133,155,0,WS_EX_TRANSPARENT
+    DEFPUSHBUTTON   "OK",IDOK,84,181,56,14
+    GROUPBOX        "USE AT YOUR OWN RISK!",IDC_STATIC,7,7,133,171,0,WS_EX_TRANSPARENT
+	CONTROL         "Preload Data Frame",IDC_PRELOAD_GS,"Button",BS_AUTOCHECKBOX | WS_TABSTOP,14,95,80,8
```
Comment (turtleli) [inline plugins/GSdx/GSdx.rc]: Use spaces instead of a tab at the start of this line.

Also can you move the "Software Mode Settings" Groupbox in the main dialog up 2 dialog units so that the spacing is more consistent.

I can merge after that's done.

Comment (ssakash) [inline plugins/GSdx/GSdx.rc]: done :+1: 

I hope you meant to decrease 2 on the y co-ordinate ? coz that's what I thought.

File: plugins/GSdx/GSSetting.cpp [hunk 11]
```diff
@@ -123,11 +123,12 @@ const char* dialog_message(int ID, bool* updateText) {
 			return "Makes textures partially or fully transparent as required by emulation. May cause unusual slowdowns for some games.";
 		case IDC_LOGZ:
 			return "Treat depth as logarithmic instead of linear. Recommended setting is on unless it causes graphical glitches.";
+		// Exclusive for Hardware Renderer	
+		case IDC_PRELOAD_GS:
+			return " Uploads GS data when rendering a new frame to reproduce some effects accurately. Fixes black screen issues in games like Primal.";
```
Comment (turtleli) [inline plugins/GSdx/GSSetting.cpp]: Oops, I missed this one. Remove the space before Uploads.

Then I'll really merge it.

Comment (ssakash) [inline plugins/GSdx/GSSetting.cpp]: fixed

File: plugins/GSdx/GSSetting.cpp [hunk 12]
```diff
@@ -123,11 +123,12 @@ const char* dialog_message(int ID, bool* updateText) {
 			return "Makes textures partially or fully transparent as required by emulation. May cause unusual slowdowns for some games.";
 		case IDC_LOGZ:
 			return "Treat depth as logarithmic instead of linear. Recommended setting is on unless it causes graphical glitches.";
+		// Exclusive for Hardware Renderer	
+		case IDC_PRELOAD_GS:
+			return "Uploads GS data when rendering a new frame to reproduce some effects accurately. Fixes black screen issues in games like Primal.";
```
Comment (turtleli) [inline plugins/GSdx/GSSetting.cpp]: ..It must seem like I'm just messing around with you. I assure you I am not.

Wasn't it supposed to say Armored Core: Last Raven? Oh, also do a trailing whitespace check.

Comment (ssakash) [inline plugins/GSdx/GSSetting.cpp]: > Wasn't it supposed to say Armored Core: Last Raven? Oh, also do a trailing whitespace check.

yeah, you're right. no clue how that got swapped up @_@ 

> ..It must seem like I'm just messing around with you. I assure you I am not.

I'm not sure what gave you that impression ! 

either way, fixed the trailing whitespace and ``IDC_PRELOAD_GS`` tooltip modification.

Comment (turtleli) [inline plugins/GSdx/GSSetting.cpp]: Well, most people don't find a few issues each time. Bad habit of mine :)

File: plugins/GSdx/GSSetting.cpp [hunk 13]
```diff
@@ -123,11 +123,12 @@ const char* dialog_message(int ID, bool* updateText) {
 			return "Makes textures partially or fully transparent as required by emulation. May cause unusual slowdowns for some games.";
 		case IDC_LOGZ:
 			return "Treat depth as logarithmic instead of linear. Recommended setting is on unless it causes graphical glitches.";
+		// Exclusive for Hardware Renderer
```
Comment (ssakash) [inline plugins/GSdx/GSSetting.cpp]: I'm not sure why I bothered to even add this useless comment ? -_-
